[GUI] Installer - Add Electrum node option

jp1ac4 opened the pull request:
This is for #1223.

For now, it's possible to edit the node's settings but not to change node type.

Remaining tasks:

- [x] Revert Cargo.toml once #1222 is merged.
- [x] Update wording as per https://github.com/wizardsardine/liana/issues/1223#issuecomment-2286483134.

jp1ac4 commented:
I made a further change to the behaviour while checking the connection. It's safer to prevent the user editing config values until the connection check completes to ensure the values don't change in the meanwhile, which could mean the connection status applies to the wrong values.

I think this was already an issue with the current installer but I guess connection checks are usually quicker with a local bitcoind and so it was less likely to cause problems.

jp1ac4 commented:
I updated the liana dependency commit.

nondiremanuel commented:
For the first page, I was proposing to take the opportunity to make the text simpler and more understandable (at least in my view):

> - I already have a node
Select this option if you already have a bitcoin node running locally or remotely. Liana will connect to it.
> - I want Liana to automatically install a Bitcoin node on my device
Liana will install a pruned (i.e. "compressed") node on your computer. You won't need to do anything except having some disk space available (~25GB required) and waiting for the initial synchronization with the network (it can take some days according to your internet connection speed).
[the two parenthesis with the estimation space and time possibly should be shown only if the network chosen is mainnet]

If there is no consensus with this, we can also leave it as you did (i.e. adding the Electrum mentions) and discuss the general wording for another time.

The second page was exactly what I had in mind. Thanks!

jp1ac4 commented:
I've updated the wording as suggested with some minor changes:
- I felt "truncated" might be more accurate than "compressed", but probably pruned already conveys this so didn't include that part.
- I added "on mainnet" after ~25 GB. We don't currently have the network selection available in the view. It should be simple to add as a follow-up commit if required and only display the additional info for mainnet.
- I used "depending on" instead of "according to" as I think it's more common this way but either should be fine.

![image](https://github.com/user-attachments/assets/43a193ba-564b-41d7-8e23-377530cf8438)

I removed "full" from the title:

![image](https://github.com/user-attachments/assets/156c8220-00f0-49c7-844a-98786d1e92f0)

pythcoiner commented:
30 GB looks more correct (signet chain included but it's only 2.2GB):
```shell
~/.liana/bitcoind ❯ sudo du datadir -sh                                                                    
29G     datadir
```

pythcoiner commented:
![image](https://github.com/user-attachments/assets/1f20bd55-686d-4186-b2fb-dbf7367ebe5d)
can we add `127.0.0.1:50001` as placeholder here? 
even if it's not likely the electrum server is on localhost, it indicate the user it should specify the port, and indicate 50001 is the default port.

pythcoiner commented:
i'm wondering about one thing here, did the address still only accept ip addresses? 
i think the default config on umbrel is to give access via `umbrel.local` instead of an ip as (i guess) almost user do not configure a static ip for their node.

pythcoiner commented on gui/src/installer/view.rs:
```diff
@@ -1088,12 +1091,107 @@ pub fn help_backup<'a>() -> Element<'a, Message> {
+pub fn define_bitcoin_node<'a>(
     progress: (usize, usize),
+    available_node_types: impl Iterator<Item = NodeType>,
+    selected_node_type: &NodeType,
```
`NodeType` implement `Copy`
```suggestion
    selected_node_type: NodeType,
```

pythcoiner commented on gui/src/installer/step/node/mod.rs:
```diff
@@ -0,0 +1,247 @@
+}
+
+impl NodeDefinition {
+    fn new(node_type: &NodeType) -> Self {
```
`NodeType` implement `Copy`
```suggestion
    fn new(node_type: NodeType) -> Self {
```

pythcoiner commented on gui/src/installer/step/node/mod.rs:
```diff
@@ -0,0 +1,247 @@
+}
+
+impl Node {
+    fn new(node_type: &NodeType) -> Self {
```
`NodeType` implement `Copy`
```suggestion
    fn new(node_type: NodeType) -> Self {
```

pythcoiner commented on gui/src/installer/view.rs:
```diff
@@ -1088,12 +1091,107 @@ pub fn help_backup<'a>() -> Element<'a, Message> {
+                }
+            })
+        } else {
+            Some(Container::new(Space::with_height(Length::Fixed(25.0))))
```
nit (maybe depending on renderer/platform)
```suggestion
            Some(Container::new(Space::with_height(Length::Fixed(21.0))))
```

pythcoiner commented:
> i'm wondering about one thing here, did the address still only accept ip addresses? i think the default config on umbrel is to give access via `umbrel.local` instead of an ip as (i guess) almost user do not configure a static ip for their node. cf [this comment](https://github.com/wizardsardine/liana/issues/1082#issuecomment-2071211975)

tested on linux adding `192.168.1.21 electrum.local` to /etc/hosts:
![image](https://github.com/user-attachments/assets/bb3fcc75-3cb5-4c95-acdc-395d9919d64c)

jp1ac4 commented on gui/src/installer/step/node/electrum.rs:
```diff
@@ -0,0 +1,73 @@
+        !self.address.value.is_empty() && self.address.valid
+    }
+
+    pub fn load_context(&mut self, _ctx: &Context) {}
```
I think it can be removed. It's a leftover from an earlier version where I defined a trait that each node type had to implement, but now it's not required.

jp1ac4 replied:
(for future reference: I removed the trait as I was having trouble including the `ping()` method in an object-safe way. I only wanted the ping method to return the `Result` and not the `Command` to handle the `PingResult`)

jp1ac4 commented on gui/src/installer/step/node/electrum.rs:
```diff
@@ -0,0 +1,73 @@
+                message::DefineElectrum::ConfigFieldEdited(field, value) => match field {
+                    ConfigField::Address => {
+                        self.address.value.clone_from(&value);
+                        self.address.valid = !self.address.value.is_empty();
```
Yes I think that would make sense, will add it.

jp1ac4 commented:
I've made the suggested changes.

I now need to rebase on master as there's a conflict.

jp1ac4 commented:
Rebased on master.

jp1ac4 commented:
Sorry, I made a further change to the Electrum address validation.

pythcoiner commented:
testing w/ several remote servers:
![image](https://github.com/user-attachments/assets/b90da073-b51c-4f2e-baa1-96a93ec2b440)
![image](https://github.com/user-attachments/assets/14de55c5-6e9b-4e42-bda8-428ccb5a311d)
![image](https://github.com/user-attachments/assets/40c71725-f5e0-48be-a5a0-1b305eadb1ac)

seems ok w/ both ssl/tcp

pythcoiner commented:
Here we should rename "Bitcoin Core" into "Electrum server":
![image](https://github.com/user-attachments/assets/4da20711-6e23-4de7-8beb-1d14abb35593)

pythcoiner commented:
Here we should also remove the rescan feature if electrum, but we should keep the block height, i often use it for know if liana is sync w/ the backend:

![image](https://github.com/user-attachments/assets/2dd056b4-ff1a-400a-a933-2554d46bbd74)

pythcoiner commented:
I'm actually trying electrum (on an other machine on my local network) on regtest w/ 550 coins:
![image](https://github.com/user-attachments/assets/48050301-6bca-4607-a651-ae3d1c572a41)

for every sync (so every 30 seconds) electrs needs around 10 seconds to answer the gui request.

if i click on `Receive` menu during this 10 seconds i got not address util all the requests done:

![image](https://github.com/user-attachments/assets/f6f545fe-36ac-4bb7-afce-77fa0e7b09eb)

i wonder about 2 things:
- do we make some call to te electrum server for generate address? or it's the iced "event loop" that became overloaded?

- why we still make so much requests to electrum server after we are sync?

pythcoiner commented:
trying to connect a mainnet wallet on regtest electrum fail as expected:
![image](https://github.com/user-attachments/assets/6e3d2b07-6403-4e49-a339-623ddbe6e2d6)
maybe we can try to detect network by genesis hash and display a more explicit message to the user?
(i don't think this should be tackle in this PR but rather in a follow up)

jp1ac4 commented:
Updated liana dependency commit hash.

jp1ac4 commented:
Added a couple of commits to update the Settings and allow to view & edit Electrum settings.

It's NOT possible to switch from one node type to another. This will require more changes and we'll probably need to leave that for a follow-up.

I've renamed the Settings page to "Node":

![image](https://github.com/user-attachments/assets/66065207-f812-4f05-bac9-d63789a14d95)

We can view the Electrum settings (and some chain info):

![image](https://github.com/user-attachments/assets/7470ccbf-bea3-4b2e-bb18-a9a636a6cd52)

We can edit the address:

![image](https://github.com/user-attachments/assets/d6d6d9f6-6bea-4f48-b772-c43e422ed7e4)

jp1ac4 commented:
TODO: Update comment in installer about rescan after importing wallet.

jp1ac4 commented:
> TODO: Update comment in installer about rescan after importing wallet.

Updated the wording.

jp1ac4 commented:
Updated liana dependency commit hash.

pythcoiner commented:
![image](https://github.com/user-attachments/assets/eb6a2252-fff9-45d9-a7b1-4fb82f87627d)
we do not ping the electrum server before let user save settings, so if the address is wrong `ElectrumSettings.processing` is still true and we cannot edit the address anymore

pythcoiner commented:
![image](https://github.com/user-attachments/assets/04a580a6-0097-44b3-a93d-a720eca9cac0)
if we conect to a wrong network electrum server we should have a clearer message

jp1ac4 commented:
Connecting to the node only after saving the changes is also done for bitcoind. Ideally, we would ping the server and check the network before saving, which could avoid that error about hash mismatch. Or perhaps I can add something to display that error differently in the GUI. Probably in a follow-up.

For Electrum, it's using a timeout of 180 seconds but I checked that it eventually gives a daemon error. I'll see if I can use a shorter timeout in #1222.

jp1ac4 commented:
> For Electrum, it's using a timeout of 180 seconds but I checked that it eventually gives a daemon error. I'll see if I can use a shorter timeout in #1222.

Hmm don't think I can change it just for the initial connection. I think better to add the pre-save ping on the GUI settings page for both bitcoind and Electrum, reusing what's already done in the installer.

jp1ac4 commented:
> The loader still refers to Bitcoin Core syncing. We can change that depending on node type in a follow-up.

Created https://github.com/wizardsardine/liana/issues/1256.

pythcoiner commented:
> > For Electrum, it's using a timeout of 180 seconds but I checked that it eventually gives a daemon error. I'll see if I can use a shorter timeout in #1222.
> 
> Hmm don't think I can change it just for the initial connection. I think better to add the pre-save ping on the GUI settings page for both bitcoind and Electrum, reusing what's already done in the installer.

maybe we can just disable edit for now (or keep the bug) and doing it in a follow up

jp1ac4 commented:
> maybe we can just disable edit for now (or keep the bug) and doing it in a follow up

I'll create a follow-up issue. I think this is the same as current behaviour for bitcoind, just that Electrum takes longer to fail.

jp1ac4 commented:
Rebased on master and updated liana dependency commit hash.

TODO: incorporate daemon changes related to rescan logic in #1222.

pythcoiner commented:
Maybe we can add (in a follow up) the name/type of the electrum server (by sending a `server.banner` request

jp1ac4 commented:
Updated liana dependency commit hash.
Changed `DaemonControl` to mutable following changes to `start_rescan` in #1222.
Allow to start rescan also for Electrum.

jp1ac4 commented:
Added a commit to enable changing from one node type to another in the settings.

Note that in its current form, it'll be possible for someone who's using managed bitcoind to switch to Electrum in the settings. In this case, the GUI will not try to start the managed bitcoind the next time they launch the GUI. I could also prevent such a user from switching node type (indeed, they cannot edit the bitcoind settings currently).

I also made a fix to a previous commit so that we take into account whether a rescan is in progress in order to determine whether the Electrum settings can be edited.

pythcoiner commented:
ACK 819eb920c0210b9d95689f40cfbeba65c03280af
